segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -196,65 +333,128 @@ public boolean supportsTruncation() {
+            }
+            if (getEpocFromPath(lastFile.getPath()) > this.context.epoch) {
+                throw new StorageNotPrimaryException(handle.getSegmentName());
+            }
+        } catch (IOException e) {
+            HDFSExceptionHelpers.throwException(handle.getSegmentName(), e);
+        }
+
+        Timer timer = new Timer();
+        try (FSDataOutputStream stream = this.context.fileSystem.append(lastFile.getPath())) {
+            if (offset != lastFile.getLen()) {
+                // Do the handle offset validation here, after we open the file. We want to throw FileNotFoundException
+                // before we throw BadOffsetException.
+                throw new BadOffsetException(handle.getSegmentName(), lastFile.getLen(), offset);
+            } else if (stream.getPos() != offset) {
+                // Looks like the filesystem changed from underneath us. This could be our bug, but it could be something else.
+                // Update our knowledge of the filesystem and throw a BadOffsetException - this should cause upstream code
+                // to try to reconcile; if it can't then the upstream code should shut down or take other appropriate measures.
+                log.warn("File changed detected for '{}'. Expected length = {}, actual length = {}.", lastFile, lastFile.getLen(), stream.getPos());
+                throw new BadOffsetException(handle.getSegmentName(), lastFile.getLen(), offset);
+            }
+
+            if (length == 0) {
+                // Exit here (vs at the beginning of the method), since we want to throw appropriate exceptions in case
+                // of Sealed or BadOffset
+                // Note: IOUtils.copyBytes with length == 0 will enter an infinite loop, hence the need for this check.
+                return;
+            }
+
+            // We need to be very careful with IOUtils.copyBytes. There are many overloads with very similar signatures.
+            // There is a difference between (InputStream, OutputStream, int, boolean) and (InputStream, OutputStream, long, boolean),
+            // in that the one with "int" uses the third arg as a buffer size, and the one with "long" uses it as the number
+            // of bytes to copy.
+            IOUtils.copyBytes(data, stream, (long) length, false);
+
+            stream.flush();
+        } catch (IOException ex) {
+            HDFSExceptionHelpers.throwException(handle.getSegmentName(), ex);
+        }
+
+        HDFSMetrics.WRITE_LATENCY.reportSuccessEvent(timer.getElapsed());
+        HDFSMetrics.WRITE_BYTES.add(length);
+        LoggerHelpers.traceLeave(log, "write", traceId, handle, offset, length);
     }
 
     @Override
-    @SneakyThrows(StreamSegmentException.class)
-    public boolean exists(String streamSegmentName) {
-        return call(new ExistsOperation(streamSegmentName, this.context));
+    public SegmentHandle openWrite(String streamSegmentName) throws StreamSegmentException {
```

Comment:
I am confused by this method. If it gets fenced, isn't that it, it should abort because someone else owns the segment? Why does it retry?

Reply:
This is to cover for the corner case where more than two hosts are contending for the segment. Consider host `A` is the current owner with epoch `1`. Now ownership changes and because of some reason hosts `B` and `C` with epochs `2` and `3` try to own the segment. It may happen that `B` wins the race to rename and `C` fails with `FileNotFound`.
With the current code `C` will go back again and own the segment.
In case `C` wins the race, and `B` gets `FileNotFound`, `B` goes back and sees that the file is now owned by a host with higher epoch and stops trying.